core/app/models/spree/promotion.rb

```diff
@@ -26,31 +31,45 @@ class Promotion < Spree::Base
 
     before_save :normalize_blank_values
 
-    scope :coupons, ->{ where("#{table_name}.code IS NOT NULL") }
-
     order_join_table = reflect_on_association(:orders).join_table
-
-    scope :applied, -> { joins("INNER JOIN #{order_join_table} ON #{order_join_table}.promotion_id = #{table_name}.id").uniq }
+    scope :applied, -> do
+      joins("INNER JOIN #{order_join_table} ON " \
+        "#{order_join_table}.promotion_id = #{table_name}.id").distinct
+    end
+    scope :coupons, -> do
+      joins(:promotion_codes).where("#{Spree::PromotionCode.table_name}.value IS NOT NULL")
+    end
 
     def self.advertised
       where(advertise: true)
     end
 
     def self.with_coupon_code(coupon_code)
-      where("lower(#{self.table_name}.code) = ?", coupon_code.strip.downcase).first
+      joins(:promotion_codes).where(
+        "lower(#{Spree::PromotionCode.table_name}.value) = ?", coupon_code.strip.downcase
+      ).first
     end
 
     def self.active
-      where('starts_at IS NULL OR starts_at < ?', Time.now).
-        where('expires_at IS NULL OR expires_at > ?', Time.now)
+      where("#{Spree::Promotion.table_name}.starts_at IS NULL OR " \
+        "#{Spree::Promotion.table_name}.starts_at < ? AND " \
+        "#{Spree::Promotion.table_name}.expires_at IS NULL OR " \
+        "#{Spree::Promotion.table_name}.expires_at > ?", Time.now, Time.now)
     end
 
     def self.order_activatable?(order)
       order && !UNACTIVATABLE_ORDER_STATES.include?(order.state)
     end
 
-    def expired?
-      !!(starts_at && Time.now < starts_at || expires_at && Time.now > expires_at)
+    def expired?(coupon_code = nil)
+      promo_exp = !!(starts_at && Time.now < starts_at || expires_at && Time.now > expires_at)
+      code      = codes.where(value: coupon_code) if coupon_code.present?
+
+      if code.present?
```

Comment:
Okay. Never mind. Not I see what's going on. Got confused briefly.

```
code      = codes.where(value: coupon_code) if coupon_code.present?
```

That should use `if coupon_code` not `.present?`  It returns an ActiveRecord::Relation.

`if code.present?` checks to see if any results are returns in the ActiveRecord::Relation (rather than just checking `if code.any?`)